gettsim/transfers/grundrente.py

```diff
@@ -0,0 +1,399 @@
+
+    Returns
+    -------
+
+    """
+    out = (
+        bonus_entgeltpunkte_grundr
+        * gr_bewertungszeiten.clip(
+            upper=ges_renten_vers_params["grundrentenzeiten"]["max"]
+        )
+        * rentenwert
+        * zugangsfaktor.clip(upper=1)
+    )
+    return out.round(2)
+
+
+def durchschnittl_entgeltpunkte_grundr(
+    entgeltpunkte_grundrente: FloatSeries, gr_bewertungszeiten: IntSeries
+) -> FloatSeries:
+    """Compute average number of Entgeltpunkte earned per month of Grundrentenzeiten.
+
+    Parameters
+    ----------
+    entgeltpunkte_grundrente
+        See basic input variable
+        :ref:`entgeltpunkte_grundrente <entgeltpunkte_grundrente>`.
+    gr_bewertungszeiten
+        See basic input variable
+        :ref:`gr_bewertungszeiten <gr_bewertungszeiten>`.
+
+    Returns
+    -------
+
+    """
+    return entgeltpunkte_grundrente / gr_bewertungszeiten
+
+
+def höchstwert_grundr_m(
+    grundrentenzeiten: IntSeries, ges_renten_vers_params: dict
+) -> FloatSeries:
+    """Calculates the maximum allowed number of average Entgeltpunkte (per month)
+    after adding bonus of Entgeltpunkte for a given number of Grundrentenzeiten.
+
+    Parameters
+    ----------
+    grundrentenzeiten
+        See basic input variable :ref:`grundrentenzeiten <grundrentenzeiten>`.
+    ges_renten_vers_params
+        See params documentation :ref:`ges_renten_vers_params <ges_renten_vers_params>`.
+
+    Returns
+    -------
+
+    """
+
+    # Calculate number of months above minimum threshold
+    months_above_thresh = (
+        grundrentenzeiten.clip(upper=ges_renten_vers_params["grundrentenzeiten"]["max"])
+        - ges_renten_vers_params["grundrentenzeiten"]["min"]
+    )
+
+    # Calculate höchstwert
+    out = (
+        ges_renten_vers_params["höchstwert_gr"]["base"]
+        + ges_renten_vers_params["höchstwert_gr"]["increment"] * months_above_thresh
+    )
+
+    # Round to 4 digits
+    return out.round(4)
+
+
+def bonus_entgeltpunkte_grundr(
+    durchschnittl_entgeltpunkte_grundr: FloatSeries,
+    höchstwert_grundr_m: FloatSeries,
+    grundrentenzeiten: IntSeries,
+    ges_renten_vers_params: dict,
+) -> FloatSeries:
+    """Calculate additional Entgeltpunkte for pensioner.
+
+    In general, the average of monthly Entgeltpunkte earnd in Grundrentenzeiten is
+    doubled, or extended to the individual Höchstwert if doubling would exceed the
+    Höchstwert. Then, the value is always multiplied by 0.875.
+
+    Parameters
+    ----------
+    durchschnittl_entgeltpunkte_grundr
+        See :func:`durchschnittl_entgeltpunkte_grundr`.
+    höchstwert_grundr_m
+        See :func:`höchstwert_grundr_m`.
+    grundrentenzeiten
+        See basic input variable :ref:`grundrentenzeiten <grundrentenzeiten>`.
+    ges_renten_vers_params
+        See params documentation :ref:`ges_renten_vers_params <ges_renten_vers_params>`.
+
+    Returns
+    -------
+
+    """
+    # Categories for distinct cases
```

Comment:
can you spell out these four distinct cases?

Reply:
Very valuable comment.. I improved the readability of this function